R/estimateEvenness.R

```diff
@@ -0,0 +1,177 @@
+#' \sQuote{Evar} and \sQuote{Bulla} evenness measures.
+#'
+#' @param x a \code{\link{SummarizedExperiment}} object
+#'
+#' @param abund_values the name of the assay used for calculation of the
+#'   sample-wise estimates
+#'
+#' @param index a evenness measurement
+#'
+#' @param name a name for the column of the colData the results should be stored
+#'   in.
+#'
+#' @param threshold a numeric threshold. assay values below or equal to this
+#'   threshold will be set to zero.
+#'
+#' @param BPPARAM A
+#'   \code{\link[BiocParallel:BiocParallelParam-class]{BiocParallelParam}}
+#'   object specifying whether calculation of estimates should be parallelized.
+#'
+#' @param ... additional parameters passed to \code{estimateEvenness}
+#'
+#' @return \code{x} with additional \code{\link{colData}} named
+#'   \code{*name*}
+#'
+#' @seealso
+#' \code{\link[scater:plotColData]{plotColData}}
+#'
+#' @name estimateEvenness
+#'
+#' @examples
+#' data(GlobalPatterns)
+#' se <- GlobalPatterns
+#'
+#' se <- estimateEvenness(se)
+#' head(colData(se))
+NULL
+
+#' @rdname estimateEvenness
+#' @export
+setGeneric("estimateEvenness",signature = c("x"),
+           function(x, abund_values = "counts",
+                    index = c("camargo", "pielou", "simpson", "evar", "bulla"),
+                    name = index, ...)
+               standardGeneric("estimateEvenness"))
+
+#' @rdname estimateEvenness
+#' @export
+setMethod("estimateEvenness", signature = c(x = "SummarizedExperiment"),
+    function(x, abund_values = "counts",
+             index = c("camargo", "pielou", "simpson", "evar", "bulla"),
+             name = index, ..., BPPARAM = SerialParam()){
+        # input check
+        index<- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                 "same length than 'index'.",
+                 call. = FALSE)
+        }
+        .check_abund_values(abund_values, x)
+        #
+        vnss <- BiocParallel::bplapply(index,
+                                       .run_evenness,
+                                       mat = assay(x, abund_values),
+                                       BPPARAM = BPPARAM, ...)
+        .add_values_to_colData(x, vnss, name)
+    }
+)
+
+.calc_bulla_evenness <- function(mat) {
+    # Species richness (number of species)
+    S <- colSums2(mat > 0, na.rm = TRUE)
+
+    # Relative abundances
+    p <- t(mat)/colSums2(mat, na.rm = TRUE)
+
+    i <- seq_len(nrow(p))
+    O <- vapply(i,function(i){sum(pmin(p[i,], 1/S[i]))},numeric(1))
+
+    # Bulla's Evenness
+    (O - 1/S)/(1 - 1/S)
+}
+
+# Camargo's eveness x: species counts zeroes: include zeros Inspired
+# by code from Pepijn de Vries and Zhou Xiang at
+# researchgate.net/post/How_can_we_calculate_the_Camargo_evenness_index_in_R
+# but rewritten here
+.calc_camargo_evenness <- function(mat) {
+    N <- colSums2(mat > 0, na.rm = TRUE)
```

Comment:
This operation is also used in several other evenness indices. Some performance gains are possible if this would be calculated before calling these functions (as they are all ran by default ie this operation gets executed multiple times).

Reply:
see below